utils/json-utils/src/main/scala/pl/touk/nussknacker/engine/json/encode/BestEffortJsonSchemaEncoder.scala

```diff
@@ -0,0 +1,90 @@
+      }
+      .toList.sequence.map { values => Json.fromFields(values) }
+  }
+
+  private def notNullOrRequired(field: (String, _), parentSchema: ObjectSchema): Boolean = {
+    field._2 != null || parentSchema.getRequiredProperties.contains(field._1)
+  }
+
+  private def encodeCollection(collection: Traversable[_], schema: ArraySchema): WithError[Json] = {
+    collection.map(el => encode(el, schema.getAllItemSchema)).toList.sequence.map(l => Json.fromValues(l))
+  }
+
+  private def encodeBasedOnSchema(value: Any, schema: Schema, fieldName: Option[String] = None): WithError[Json] = {
+    (schema, value) match {
+      case (schema: ObjectSchema, map: scala.collection.Map[String@unchecked, _]) => encodeObject(map.toMap, schema)
+      case (schema: ObjectSchema, map: java.util.Map[String@unchecked, _]) => encodeObject(map.toMap, schema)
+      case (schema: ArraySchema, value: Traversable[_]) => encodeCollection(value, schema)
+      case (schema: ArraySchema, value: java.util.Collection[_]) => encodeCollection(value.toArray, schema)
+      case (_: StringSchema, value: String) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Long) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Double) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Float) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Int) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: java.math.BigDecimal) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: java.math.BigInteger) => Valid(jsonEncoder.encode(value))
+      case (_: NumberSchema, value: Number) => Valid(jsonEncoder.encode(value.doubleValue()))
+      // todo dates
+      //      case a: LocalDate => Encoder[LocalDate].apply(a)
+      //      case a: LocalTime => Encoder[LocalTime].apply(a)
+      //      case a: LocalDateTime => Encoder[LocalDateTime].apply(a)
+      //      Default implementation serializes to ISO_ZONED_DATE_TIME which is not handled well by some parsers...
+      //      case a: ZonedDateTime => encodeZonedDateTimeWithFormatter(DateTimeFormatter.ISO_OFFSET_DATE_TIME).apply(a)
+      //      case a: Instant => Encoder[Instant].apply(a)
+      //      case a: OffsetDateTime => Encoder[OffsetDateTime].apply(a)
+      //      case a: UUID => safeString(a.toString)
+      //      case a: DisplayJson => a.asJson
+      case (_: NullSchema, null) => Valid(Json.Null)
+      case (_: NullSchema, None) => Valid(Json.Null)
+      case (_: BooleanSchema, value: Boolean) => Valid(Json.fromBoolean(value))
+      case (_: EnumSchema, value: Enum[_]) => Valid(Json.fromString(value.toString))
+      case (null, value: Any) if validationMode == ValidationMode.lax => Valid(Json.fromString(value.toString))
```

Comment:
schema can be `null`? Is it possible, and if so - is it ok?

Reply:
Exactly - like we talk before, we shouldn't check validation mode but additionalProperties (which is validation mode  inside schema) We don't want to mix both things

Reply:
We can do this things incrementally. I just want to be sure that we never produce invalid messages. Currently for validationMode = lax and additionalProperties = false will be some exception in runtime? If not, I'm for adding throwing exception if result is invalid.